# Data Mining Project - Group XX 2025/2026

# Import Libraries

In [83]:
import sqlite3
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from math import ceil

from itertools import product
from ydata_profiling import ProfileReport

# for better resolution plots
%config InlineBackend.figure_format = 'retina'

#o svg consegue ampliar infinitamente os gráficos sem perder qualidade mas às vezes é mais lento 
#por isso agora usamos retina


sns.set()

# Data Exploration and Initial Analysis

## Loading the data

Import the datasets from csv files using commas as separators of the columns and setting the unique customer identifier as the index of both columns.

In [84]:
flightsDB = pd.read_csv('DM_AIAI_FlightsDB.csv', sep = ",", index_col= "Loyalty#")
customerDB = pd.read_csv('DM_AIAI_CustomerDB.csv', sep = ",", index_col= "Loyalty#")
metaData = pd.read_csv('DM_AIAI_Metadata.csv', sep = ";", header= None)

Remove the 'Unnamed' column referring to a sequential numbering of the rows, as we set the column "Loyalty#" as the index

In [85]:
customerDB = customerDB.iloc[:, 1:]
customerDB

First Name    Last Name        Customer Name Country  \
Loyalty#                                                        
480934      Cecilia  Householder  Cecilia Householder  Canada   
549612        Dayle        Menez          Dayle Menez  Canada   
429460       Necole       Hannon        Necole Hannon  Canada   
608370        Queen        Hagee          Queen Hagee  Canada   
530508       Claire      Latting       Claire Latting  Canada   
...             ...          ...                  ...     ...   
100012        Ethan     Thompson       Ethan Thompson  Canada   
100013        Layla        Young          Layla Young  Canada   
100014       Amelia      Bennett       Amelia Bennett  Canada   
100015     Benjamin       Wilson      Benjamin Wilson  Canada   
100016         Emma       Martin          Emma Martin  Canada   

         Province or State          City   Latitude   Longitude Postal code  \
Loyalty#                                                                      
480934             Ontario       Toronto  43.653225  -79.383186     M2Z 4K1   
549612             Alberta      Edmonton  53.544388 -113.490930     T3G 6Y6   
429460    British Columbia     Vancouver  49.282730 -123.120740     V6E 3D9   
608370             Ontario       Toronto  43.653225  -79.383186     P1W 1K4   
530508              Quebec          Hull  45.428730  -75.713364     J8Y 3Z5   
...                    ...           ...        ...         ...         ...   
100012              Quebec   Quebec City  46.759733  -71.141009     Y0C 7D6   
100013             Alberta      Edmonton  53.524829 -113.546357     L3S 9Y3   
100014       New Brunswick       Moncton  46.051866  -64.825428     G2S 2B6   
100015              Quebec   Quebec City  46.862970  -71.133444     B1Z 8T3   
100016    British Columbia  Dawson Creek  55.720562 -120.160090     M4A 1E4   

          Gender Education Location Code   Income Marital Status  \
Loyalty#                                                           
480934    female  Bachelor         Urban  70146.0        Married   
549612      male   College         Rural      0.0       Divorced   
429460      male   College         Urban      0.0         Single   
608370      male   College      Suburban      0.0         Single   
530508      male  Bachelor      Suburban  97832.0        Married   
...          ...       ...           ...      ...            ...   
100012      male  Bachelor      Suburban      NaN         Single   
100013    female  Bachelor         Rural      NaN        Married   
100014      male  Bachelor         Rural      NaN        Married   
100015    female   College         Urban      NaN        Married   
100016    female    Master      Suburban      NaN         Single   

         LoyaltyStatus EnrollmentDateOpening CancellationDate  \
Loyalty#                                                        
480934            Star             2/15/2019              NaN   
549612            Star              3/9/2019              NaN   
429460            Star             7/14/2017         1/8/2021   
608370            Star             2/17/2016              NaN   
530508            Star            10/25/2017              NaN   
...                ...                   ...              ...   
100012            Star             2/27/2019        2/27/2019   
100013            Star             9/20/2017        9/20/2017   
100014            Star            11/28/2020       11/28/2020   
100015            Star              4/9/2020         4/9/2020   
100016            Star             7/21/2020        7/21/2020   

          Customer Lifetime Value  EnrollmentType  
Loyalty#                                           
480934                    3839.14        Standard  
549612                    3839.61        Standard  
429460                    3839.75        Standard  
608370                    3839.75        Standard  
530508                    3842.79  2021 Promotion  
...                           ...       

## Metadata

In [86]:
# display(metaData)

**CustomerDB Database Variable Description**
- **Loyalty#:**  Unique customer identifier for loyalty program members
- **First Name:**   Customer's first name
- **Last Name:**   Customer's last name 
- **Customer Name:** Customer's full name (concatenated)
- **Country:**	Customer's country of residence
- **Province or State:**	Customer's province or state
- **City:**	Customer's city of residence
- **Latitude:**	Geographic latitude coordinate of customer location
- **Longitude:**	Geographic longitude coordinate of customer locatio
- **Postal code:**	Customer's postal/ZIP code
- **Gender:**	Customer's gender
- **Education:**	Customer's highest education level (Bachelor, College, etc.)
- **Location:** Code	Urban/Suburban/Rural classification of customer residence
- **Income:**	Customer's annual income
- **Marital Status:**	Customer's marital status (Married, Single, Divorced)
- **LoyaltyStatus:**	Current tier status in loyalty program (Star > Nova > Aurora)
- **EnrollmentDateOpening:**	Date when customer joined the loyalty program
- **CancellationDate:**	Date when customer left the program
- **Customer Lifetime:** Value	Total calculated monetary value of customer relationship
- **EnrollmentType:**	Method of joining loyalty program


**FlightsDB Database Variable Description**
- **Loyalty#:**	Unique customer identifier linking to CustomerDB
- **Year:**	Year of flight activity record
- **Month:**	Month of flight activity record (1-12)
- **YearMonthDate:**	First day of the month for the activity period
- **NumFlights:**	Total number of flights taken by customer in the month
- **NumFlightsWithCompanions:**	Number of flights where customer traveled with companions
- **DistanceKM:**	Total distance traveled in kilometers for the month
- **PointsAccumulated:**	Loyalty points earned by customer during the month
- **PointsRedeemed:**	Loyalty points spent/redeemed by customer during the month
- **DollarCostPointsRedeemed:**	Dollar value of points redeemed during the month

# Data Understanding

On this section we will inspect the data shape, column names and data types for each dataset

## General Look at the DataSet (FlightsDB)  - Maria

In [87]:
flightsDB.shape

(608436, 9)

In [88]:
flightsDB.head(15)


Year  Month YearMonthDate  NumFlights  NumFlightsWithCompanions  \
Loyalty#                                                                    
413052    2021     12     12/1/2021         2.0                       2.0   
464105    2021     12     12/1/2021         0.0                       0.0   
681785    2021     12     12/1/2021        10.0                       3.0   
185013    2021     12     12/1/2021        16.0                       4.0   
216596    2021     12     12/1/2021         9.0                       0.0   
486956    2021     12     12/1/2021        12.0                       7.0   
247514    2021     12     12/1/2021        17.0                       7.0   
711864    2021     12     12/1/2021         6.0                       0.0   
721372    2021     12     12/1/2021        11.0                       3.0   
762715    2021     12     12/1/2021         0.0                       0.0   
332716    2021     12     12/1/2021         0.0                       0.0   
287254    2021     12     12/1/2021         7.0                       0.0   
904920    2021     12     12/1/2021         0.0                       0.0   
436971    2021     12     12/1/2021         2.0                       2.0   
671534    2021     12     12/1/2021         0.0                       0.0   

          DistanceKM  PointsAccumulated  PointsRedeemed  \
Loyalty#                                                  
413052        9384.0              938.0             0.0   
464105           0.0                0.0             0.0   
681785       14745.0             1474.0             0.0   
185013       26311.0             2631.0          3213.0   
216596       19275.0             1927.0             0.0   
486956       23967.0             2396.0             0.0   
247514       23029.0             2302.0             0.0   
711864       25995.0             2599.0             0.0   
721372       30758.0             3075.0             0.0   
762715           0.0                0.0             0.0   
332716           0.0                0.0             0.0   
287254       13328.0             1332.0             0.0   
904920           0.0                0.0             0.0   
436971       23295.0             2329.0             0.0   
671534           0.0                0.0             0.0   

          DollarCostPointsRedeemed  
Loyalty#                            
413052                         0.0  
464105                         0.0  
681785                         0.0  
185013                        32.0  
216596                         0.0  
486956                         0.0  
247514                         0.0  
711864                         0.0  
721372                         0.0  
762715                         0.0  
332716                         0.0  
287254                         0.0  
904920                         0.0  
436971                         0.0  
671534                         0.0

In [89]:
flightsDB.tail(15)

Year  Month YearMonthDate  NumFlights  NumFlightsWithCompanions  \
Loyalty#                                                                    
999498    2019     12     12/1/2019         0.9                       0.9   
999513    2019     12     12/1/2019         0.0                       0.0   
999524    2019     12     12/1/2019        13.5                       4.5   
999550    2019     12     12/1/2019         8.1                       0.0   
999589    2019     12     12/1/2019         0.0                       0.0   
999631    2019     12     12/1/2019         3.6                       1.8   
999731    2019     12     12/1/2019         0.0                       0.0   
999758    2019     12     12/1/2019         0.0                       0.0   
999788    2019     12     12/1/2019         0.0                       0.0   
999891    2019     12     12/1/2019         0.0                       0.0   
999902    2019     12     12/1/2019         7.2                       0.0   
999911    2019     12     12/1/2019         0.0                       0.0   
999940    2019     12     12/1/2019        14.4                       0.9   
999982    2019     12     12/1/2019         0.0                       0.0   
999986    2019     12     12/1/2019         0.0                       0.0   

          DistanceKM  PointsAccumulated  PointsRedeemed  \
Loyalty#                                                  
999498       30283.2            3028.32             0.0   
999513           0.0               0.00             0.0   
999524       22572.9            2257.29             0.0   
999550       18168.3            1816.83             0.0   
999589           0.0               0.00             0.0   
999631       12262.5            1226.25             0.0   
999731           0.0               0.00             0.0   
999758           0.0               0.00             0.0   
999788           0.0               0.00             0.0   
999891           0.0               0.00             0.0   
999902       30766.5            3076.65             0.0   
999911           0.0               0.00             0.0   
999940       18261.0            1826.10             0.0   
999982           0.0               0.00             0.0   
999986           0.0               0.00             0.0   

          DollarCostPointsRedeemed  
Loyalty#                            
999498                         0.0  
999513                         0.0  
999524                         0.0  
999550                         0.0  
999589                         0.0  
999631                         0.0  
999731                         0.0  
999758                         0.0  
999788                         0.0  
999891                         0.0  
999902                         0.0  
999911                         0.0  
999940                         0.0  
999982                         0.0  
999986                         0.0

From the visualization of the head and tail of the data base we can already understand that some errors exist:

    - NumFlights and NumFlightsWithCompanions as floats...
    - PointsAccumulated and PointsRedeemed as floats. Should they be integers?
We will further analyse this using describe and info.


In [90]:
flightsDB.info()

<class 'pandas.core.frame.DataFrame'>
Index: 608436 entries, 413052 to 999986
Data columns (total 9 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Year                      608436 non-null  int64  
 1   Month                     608436 non-null  int64  
 2   YearMonthDate             608436 non-null  object 
 3   NumFlights                608436 non-null  float64
 4   NumFlightsWithCompanions  608436 non-null  float64
 5   DistanceKM                608436 non-null  float64
 6   PointsAccumulated         608436 non-null  float64
 7   PointsRedeemed            608436 non-null  float64
 8   DollarCostPointsRedeemed  608436 non-null  float64
dtypes: float64(6), int64(2), object(1)
memory usage: 46.4+ MB


From info we can see that:

    - NumFlights and NumFlightsWithCompanions as floats...
    - PointsAccumulated and PointsRedeemed as floats. Should they be integers? 
    - There aren't missing values

What will we do?

    Analyse with describe to have a different view

In [91]:
#To confirm that missing values don't exist
flightsDB.replace("", np.nan, inplace=True)
flightsDB.isna().sum()

Year                        0
Month                       0
YearMonthDate               0
NumFlights                  0
NumFlightsWithCompanions    0
DistanceKM                  0
PointsAccumulated           0
PointsRedeemed              0
DollarCostPointsRedeemed    0
dtype: int64

In [92]:
flightsDB.describe().T

count         mean           std     min  \
Year                      608436.0  2020.000000      0.816497  2019.0   
Month                     608436.0     6.500000      3.452055     1.0   
NumFlights                608436.0     3.908107      5.057889     0.0   
NumFlightsWithCompanions  608436.0     0.983944      2.003785     0.0   
DistanceKM                608436.0  7939.341419  10260.421873     0.0   
PointsAccumulated         608436.0   793.777781   1025.918521     0.0   
PointsRedeemed            608436.0   235.251678    983.233374     0.0   
DollarCostPointsRedeemed  608436.0     2.324835      9.725168     0.0   

                              25%       50%         75%      max  
Year                      2019.00  2020.000   2021.0000   2021.0  
Month                        3.75     6.500      9.2500     12.0  
NumFlights                   0.00     0.000      7.2000     21.0  
NumFlightsWithCompanions     0.00     0.000      0.9000     11.0  
DistanceKM                   0.00   856.400  15338.1750  42040.0  
PointsAccumulated            0.00    85.275   1533.7125   4204.0  
PointsRedeemed               0.00     0.000      0.0000   7496.0  
DollarCostPointsRedeemed     0.00     0.000      0.0000     74.0

In [93]:
flightsDB.describe(include='object')

#o "top" é a moda e "freq" é a frequencia do valor mais frequente
#"unique" é a quantidade de valores unicos ((36 datas diferentes pq é o primeiro dia de cada mês durante 3 anos))
#"count" é o numero de valores nao nulos

YearMonthDate
count         608436
unique            36
top        12/1/2021
freq           16901

From both numeric and categorical describe we don't notice any weird value.

In [94]:
# flightsDB.dtypes

#### Check Duplicate Values

In [95]:
#Check how many duplicates exist
flightsDB.duplicated().sum()

np.int64(301411)

In [96]:
#Visualize our duplicates
flightsDB[flightsDB.duplicated()]

Year  Month YearMonthDate  NumFlights  NumFlightsWithCompanions  \
Loyalty#                                                                    
762715    2021     12     12/1/2021         0.0                       0.0   
332716    2021     12     12/1/2021         0.0                       0.0   
904920    2021     12     12/1/2021         0.0                       0.0   
671534    2021     12     12/1/2021         0.0                       0.0   
618871    2021     12     12/1/2021         0.0                       0.0   
...        ...    ...           ...         ...                       ...   
999788    2019     12     12/1/2019         0.0                       0.0   
999891    2019     12     12/1/2019         0.0                       0.0   
999911    2019     12     12/1/2019         0.0                       0.0   
999982    2019     12     12/1/2019         0.0                       0.0   
999986    2019     12     12/1/2019         0.0                       0.0   

          DistanceKM  PointsAccumulated  PointsRedeemed  \
Loyalty#                                                  
762715           0.0                0.0             0.0   
332716           0.0                0.0             0.0   
904920           0.0                0.0             0.0   
671534           0.0                0.0             0.0   
618871           0.0                0.0             0.0   
...              ...                ...             ...   
999788           0.0                0.0             0.0   
999891           0.0                0.0             0.0   
999911           0.0                0.0             0.0   
999982           0.0                0.0             0.0   
999986           0.0                0.0             0.0   

          DollarCostPointsRedeemed  
Loyalty#                            
762715                         0.0  
332716                         0.0  
904920                         0.0  
671534                         0.0  
618871                         0.0  
...                            ...  
999788                         0.0  
999891                         0.0  
999911                         0.0  
999982                         0.0  
999986                         0.0  

[301411 rows x 9 columns]

From this visualization we understand that all the rows considered as duplicates have one unique different value for the Loyalty# feature.

In [97]:
#check the percentage of duplicates in our DataFrame
flightsDB.duplicated().sum() / len(flightsDB) * 100

np.float64(49.53865320263758)

!!!!    The percentage of duplicates ir almost 50%    !!!!

Because of this we understand that having Loyalty# as an index can be a wrong approach to check the duplicates so we create a new variable with the Loyalty# as a feature to check again the duplicates considering this feature.

In [120]:
flightsDB_index = pd.read_csv('DM_AIAI_FlightsDB.csv', sep = ",")
flightsDB_index.duplicated().sum() / len(flightsDB_index) * 100

np.float64(0.4771249564457067)

From the new calculation we obtain only 0.48% of duplicated which it makes more sense in our problem.

With this value we can decide to drop the duplicates 

In [121]:
flightsDB_index[flightsDB_index["Loyalty#"] == 263267]
#Here we check that there are duplicates for the Loyalty# number 263267
#the DataFrame below show us all the Data associated to this Loyalty number and we can see that some rows have the exactly same information


Loyalty#  Year  Month YearMonthDate  NumFlights  \
1092      263267  2020      6      6/1/2020         0.0   
3150      263267  2020      6      6/1/2020         0.0   
14057     263267  2020      5      5/1/2020         0.0   
25441     263267  2020      5      5/1/2020         0.0   
37425     263267  2020      4      4/1/2020         0.0   
...          ...   ...    ...           ...         ...   
560775    263267  2019     10     10/1/2019         0.0   
577675    263267  2019     11     11/1/2019         0.0   
577676    263267  2019     11     11/1/2019         0.0   
594576    263267  2019     12     12/1/2019         0.0   
594577    263267  2019     12     12/1/2019         0.0   

        NumFlightsWithCompanions  DistanceKM  PointsAccumulated  \
1092                         0.0         0.0                0.0   
3150                         0.0         0.0                0.0   
14057                        0.0         0.0                0.0   
25441                        0.0         0.0                0.0   
37425                        0.0         0.0                0.0   
...                          ...         ...                ...   
560775                       0.0         0.0                0.0   
577675                       0.0         0.0                0.0   
577676                       0.0         0.0                0.0   
594576                       0.0         0.0                0.0   
594577                       0.0         0.0                0.0   

        PointsRedeemed  DollarCostPointsRedeemed  
1092               0.0                       0.0  
3150               0.0                       0.0  
14057              0.0                       0.0  
25441              0.0                       0.0  
37425              0.0                       0.0  
...                ...                       ...  
560775             0.0                       0.0  
577675             0.0                       0.0  
577676             0.0                       0.0  
594576             0.0                       0.0  
594577             0.0                       0.0  

[72 rows x 10 columns]

There are 72 equal rows meaning all 36 unique values (corresponding to 12 months over 3 years) are duplicated.

## General Look at the Data (CustomerDB) - Maria

In [100]:
customerDB.shape

(16921, 19)

In [101]:
customerDB.head(10)

First Name    Last Name        Customer Name Country  \
Loyalty#                                                        
480934      Cecilia  Householder  Cecilia Householder  Canada   
549612        Dayle        Menez          Dayle Menez  Canada   
429460       Necole       Hannon        Necole Hannon  Canada   
608370        Queen        Hagee          Queen Hagee  Canada   
530508       Claire      Latting       Claire Latting  Canada   
193662     Leatrice       Hanlin      Leatrice Hanlin  Canada   
927943          Hue      Sellner          Hue Sellner  Canada   
188893        Nakia         Cash           Nakia Cash  Canada   
852392       Arlene    Conterras     Arlene Conterras  Canada   
866307       Dustin       Recine        Dustin Recine  Canada   

         Province or State        City   Latitude   Longitude Postal code  \
Loyalty#                                                                    
480934             Ontario     Toronto  43.653225  -79.383186     M2Z 4K1   
549612             Alberta    Edmonton  53.544388 -113.490930     T3G 6Y6   
429460    British Columbia   Vancouver  49.282730 -123.120740     V6E 3D9   
608370             Ontario     Toronto  43.653225  -79.383186     P1W 1K4   
530508              Quebec        Hull  45.428730  -75.713364     J8Y 3Z5   
193662               Yukon  Whitehorse  60.721188 -135.056840     Y2K 6R0   
927943             Ontario     Toronto  43.653225  -79.383186     P5S 6R4   
188893             Ontario     Trenton  44.101128  -77.576309     K8V 4B2   
852392              Quebec    Montreal  45.501690  -73.567253     H2Y 2W2   
866307             Ontario     Toronto  43.653225  -79.383186     M8Y 4K8   

          Gender Education Location Code   Income Marital Status  \
Loyalty#                                                           
480934    female  Bachelor         Urban  70146.0        Married   
549612      male   College         Rural      0.0       Divorced   
429460      male   College         Urban      0.0         Single   
608370      male   College      Suburban      0.0         Single   
530508      male  Bachelor      Suburban  97832.0        Married   
193662      male  Bachelor         Rural  26262.0        Married   
927943    female   College         Urban      0.0         Single   
188893      male  Bachelor      Suburban  93272.0        Married   
852392    female  Bachelor      Suburban  93272.0        Married   
866307      male  Bachelor      Suburban  93272.0        Married   

         LoyaltyStatus EnrollmentDateOpening CancellationDate  \
Loyalty#                                                        
480934            Star             2/15/2019              NaN   
549612            Star              3/9/2019              NaN   
429460            Star             7/14/2017         1/8/2021   
608370            Star             2/17/2016              NaN   
530508            Star            10/25/2017              NaN   
193662            Star              5/7/2015              NaN   
927943            Star              6/9/2017              NaN   
188893            Star             12/8/2019              NaN   
852392            Star             5/30/2018              NaN   
866307            Star            10/14/2019              NaN   

          Customer Lifetime Value  EnrollmentType  
Loyalty#                                           
480934                    3839.14        Standard  
549612                    3839.61        Standard  
429460                    3839.75        Standard  
608370                    3839.75        Standard  
530508                    3842.79  2021 Promotion  
193662                    3844.57        Standard  
927943                    3857.95        Standard  
188893                    3861.49        Standard  
852392                    3861.49        Standard  
866307                    3861.49        Standard

In [102]:
customerDB.tail(20)

First Name Last Name    Customer Name Country Province or State  \
Loyalty#                                                                   
999987        Layla    Murphy     Layla Murphy  Canada     New Brunswick   
999988         Jana    Parker      Jana Parker  Canada            Quebec   
999989        Ethan    Parker     Ethan Parker  Canada           Ontario   
999990         Ryan  Anderson    Ryan Anderson  Canada     New Brunswick   
999991       Olivia      Cote      Olivia Cote  Canada     New Brunswick   
999992         Ella       Roy         Ella Roy  Canada           Ontario   
999993       Elijah      Cook      Elijah Cook  Canada  British Columbia   
999994        Ethan      Chan       Ethan Chan  Canada           Ontario   
999995         Liam      Wong        Liam Wong  Canada           Ontario   
999996     Isabella      Ross    Isabella Ross  Canada           Ontario   
999997        Grace    Miller     Grace Miller  Canada     New Brunswick   
999998         Noah  Thompson    Noah Thompson  Canada            Quebec   
999999         Noah    Walker      Noah Walker  Canada           Alberta   
100017        Grace    Fortin     Grace Fortin  Canada  British Columbia   
100011       Amelia      Ross      Amelia Ross  Canada           Ontario   
100012        Ethan  Thompson   Ethan Thompson  Canada            Quebec   
100013        Layla     Young      Layla Young  Canada           Alberta   
100014       Amelia   Bennett   Amelia Bennett  Canada     New Brunswick   
100015     Benjamin    Wilson  Benjamin Wilson  Canada            Quebec   
100016         Emma    Martin      Emma Martin  Canada  British Columbia   

                  City   Latitude   Longitude Postal code  Gender Education  \
Loyalty#                                                                      
999987     Fredericton  46.029263  -66.565150     R4H 2Y2  female  Bachelor   
999988        Montreal  45.573672  -73.523012     N6B 1N3    male   College   
999989         Trenton  44.075379  -77.550375     P8F 5C8    male   College   
999990         Moncton  46.106617  -64.714267     B6P 6D0  female   College   
999991     Fredericton  45.950000  -66.652437     X3W 5N2  female   College   
999992         Toronto  43.706878  -79.437412     P6D 6N2    male   College   
999993    Dawson Creek  55.701475 -120.181716     W6H 0Z7  female   College   
999994          Ottawa  45.365906  -75.723181     B2F 3E1  female   College   
999995          Ottawa  45.471557  -75.704868     B3A 2R0  female   College   
999996         Toronto  43.690489  -79.436758     B4W 4M6  female  Bachelor   
999997         Moncton  46.067243  -64.729480     S5R 0X5    male   College   
999998     Quebec City  46.749897  -71.238050     K7T 9I9    male  Bachelor   
999999        Edmonton  53.549353 -113.509109     S7D 0V1    male  Bachelor   
100017    Dawson Creek  55.751178 -120.264920     E0K 5I2    male    Master   
100011         Toronto  43.593187  -79.444335     W9D 4Q9  female  Bachelor   
100012     Quebec City  46.759733  -71.141009     Y0C 7D6    male  Bachelor   
100013        Edmonton  53.524829 -113.546357     L3S 9Y3  female  Bachelor   
100014         Moncton  46.051866  -64.825428     G2S 2B6    male  Bachelor   
100015     Quebec City  46.862970  -71.133444     B1Z 8T3  female   College   
100016    Dawson Creek  55.720562 -120.160090     M4A 1E4  female    Master   

         Location Code  Income Marital Status LoyaltyStatus  \
Loyalty#                                                      
999987           Urban     NaN         Single          Star   
999988           Rural     NaN         Single          Star   
999989           Rural     NaN        Married          Star   
999990           Rural     NaN        Married          Star   
999991        Suburban     NaN        Married          Star   
999992        Suburban     NaN         Single          Star   
999993        Suburban     NaN        Married          Star   
999994           Rural     NaN

From the visualization of the head and tail of the data base we can already understand that some errors exist:

    - A column named 'Unnamed' as an index with the number of each row
    - Missing values in some features
    - EnrollmentType as "2021 Promotion" when it's suppose to be a type
We will further analyse this using describe and info.

It's also possible to see that some variables are redundante, such as Costumer Name, First Name and Last Name
To solve this problem we will uniformize all the values in data preparation

In [103]:
customerDB.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16921 entries, 480934 to 100016
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   First Name               16921 non-null  object 
 1   Last Name                16921 non-null  object 
 2   Customer Name            16921 non-null  object 
 3   Country                  16921 non-null  object 
 4   Province or State        16921 non-null  object 
 5   City                     16921 non-null  object 
 6   Latitude                 16921 non-null  float64
 7   Longitude                16921 non-null  float64
 8   Postal code              16921 non-null  object 
 9   Gender                   16921 non-null  object 
 10  Education                16921 non-null  object 
 11  Location Code            16921 non-null  object 
 12  Income                   16901 non-null  float64
 13  Marital Status           16921 non-null  object 
 14  LoyaltyStatus        

From info we can see that:

    - missing values in Income, CustomerLifetimeValue, CancellationDate
* the missing values in the features Income can make sense in cases where customers do not want to share their personal annual income. Or they may also be input errors. (Depends on interpretation).

* We can also believe that it makes sense to have NaN values in “CancellationDate,” as this means that there are customers who have not left the program.

* For the “CustomerLifetimeValue” variable, we believe that it does not make sense to have NaN values because even if the customer has no value for the company, their CustomerLifetimeValue will be 0.

What will we do?

    Analyse with describe to have a different view

In [104]:
#To confirm that missing values exist
customerDB.replace("", np.nan, inplace=True)
customerDB.isna().sum()

First Name                     0
Last Name                      0
Customer Name                  0
Country                        0
Province or State              0
City                           0
Latitude                       0
Longitude                      0
Postal code                    0
Gender                         0
Education                      0
Location Code                  0
Income                        20
Marital Status                 0
LoyaltyStatus                  0
EnrollmentDateOpening          0
CancellationDate           14611
Customer Lifetime Value       20
EnrollmentType                 0
dtype: int64

In [105]:
customerDB.describe()

Latitude     Longitude        Income  Customer Lifetime Value
count  16921.000000  16921.000000  16901.000000             16901.000000
mean      47.174500    -91.814768  37758.038400              7990.460188
std        3.307971     22.242429  30368.992499              6863.173093
min       42.984924   -135.056840      0.000000              1898.010000
25%       44.231171   -120.237660      0.000000              3979.720000
50%       46.087818    -79.383186  34161.000000              5780.180000
75%       49.282730    -74.596184  62396.000000              8945.690000
max       60.721188    -52.712578  99981.000000             83325.380000

From the numerical describe we can see that:

    - Once again we have the column Unnamed that has no relevant values

From the rest of the infromation we can't find any other problem from the first look

In [106]:
customerDB.describe(include='object')

First Name Last Name        Customer Name Country Province or State  \
count       16921     16921                16921   16921             16921   
unique       4941     15404                16921       1                11   
top          Deon   Salberg  Cecilia Householder  Canada           Ontario   
freq           13         4                    1   16921              5468   

           City Postal code  Gender Education Location Code Marital Status  \
count     16921       16921   16921     16921         16921          16921   
unique       29          75       2         5             3              3   
top     Toronto     V6E 3D9  female  Bachelor      Suburban        Married   
freq       3390         917    8497     10586          5716           9842   

       LoyaltyStatus EnrollmentDateOpening CancellationDate EnrollmentType  
count          16921                 16921             2310          16921  
unique             3                  2449             1260              2  
top             Star              4/3/2015         7/7/2020       Standard  
freq            7761                    34                8          15773

From the object describe we can conclude that:

    - there are no repeted Customer Names (count = unique = 16921);
    - there's only one Country, Canada
    - other things that will be analysed latter if they are relevant

#### Check Duplicates

In [107]:
customerDB.duplicated().sum()

np.int64(0)

Checking the duplicates we verify that we don´t have any.

But it's still important to check the duplicates without the names features.

In [108]:
customerDB_no_name = customerDB.drop(columns=["First Name", "Last Name", "Customer Name"])
customerDB_no_name.duplicated().sum()

np.int64(0)

The result is the same so we can conclude that there aren't duplicated values in this DataFrame.

# Data Quality Check - Maria e Margarida

To do on this section:
- Identifying missing values - **Margarida**
- Checking and correcting data types - **Maria** (general look at the data)
- detecting and handling duplicate records (handling duplicates?? não é só detecting?) - **Maria** (general look at the data)

# Data Agregation and Exploration - Maria

To do on this section:
- Summing and agregating  data by columns and rows
- Discussing the appropriateness of different operations

## Falta fazer:
1. Feature Engineering - Maria
2. Identify Strange Values - Lourenço